December review (#162)
hassanobeid1994/tr_b_causal_2020: article/sections/_6_graph_discovery.tex

Similarly, undirected or bi-directed edges between variables highlight causal ambiguity.
Such call-outs invite analysts to resolve the question of directionality in the relationship.
And conversely, we may learn from causal links (i.e. arrows) that are present in the graphs output from our causal discovery algorithm that we initially overlooked.
By contrasting and criticizing alternative graphs, we clarify the strengths and deficiencies of our own point of view.
Then, once we've identified elements that we think should or should not be present in a causal graph for our dataset, we can amend our hand-crafted graph to meet these requirements.

% How does causal discovery help characterize posterior uncertainty about the  true causal graph?
% Weighted bootstrap posterior
Lastly, causal discovery can help us characterize our posterior uncertainty about the data-generating causal graph.
They enable approximation of the posterior distribution over causal graphs, in at least two ways.
One approach is to use a weighted likelihood bootstrap approximation \citep{newton_1994_approximate} to the posterior distribution over graphs.

> In this approach, one would first sample a vector of weights for the likelihood terms.

Author: ```suggestion
In this approach, one would first sample a vector of sample weights for the likelihood terms.
```

Author: See edits in #170.

Then, one would use those weights in one's causal discovery algorithm to produce a single `sample' from the posterior approximation.

%Randomize-then-optimize.
Alternatively, we could use yet another randomize-then-optimize \citep{bardsley_2014_randomize, orabona_2014_measure} approach to sampling from an approximate posterior distribution.
Here, one would sample from a prior on entries of the matrix representation of a causal graph.
Semantically, we sample constraints such as ``$X \rightarrow Y$ (MUST | MUST NOT) be in the causal graph''.
We ``sample'' from the approximate posterior by running the causal discovery algorithm on the original dataset, with the inclusion of these randomly generated constraints.
And, of course, we can consider hybrids of these two posterior approximation schemes.

With these posterior approximation methods, we can generate causal graphs for all the purposes mentioned above:
\begin{itemize}
   \item for criticism and inspiration,
   \item for distributional analyses of the causal graphs, and

>    \item for distributional analyses of one's causal effect estimates conditional on each sampled graph.\newline

Author: ```suggestion
   \item for distributional analyses of one's causal effect estimates, conditional on each sampled graph.\newline
```

Author: See edits in #170.

And of course, extensions of these techniques exist for inference in the presence of unobserved confounding \citep[Sec. 6]{goudet_2018_learning} and for learning an entire causal graph as opposed to dealing solely with variable pairs \citep{zheng_2020_learning}.

\subsection{An application of causal discovery}
\label{sec:discovery-application}
To demonstrate the methods in this section, we chose to use the simplest causal discovery algorithm: the PC algorithm \citep{glymour_2001_causation}.
For practitioners who may reasonably start with the simplest method available and increase methodological complexity as needed, this should be an illuminating starting point.
As noted in Section \ref{sec:testing-demonstration}, we suspect that our causal graph for the drive alone utility of our dataset contains unobserved confounding.
Given that the PC algorithm assumes that we are free of unobserved confounding, our example allows us to observe how the algorithm behaves under violation of its assumptions.
Does the algorithm fail gracefully and point towards violations of its assumptions, or does it return incorrect results with certainty?

> To begin, it helps to understand the general procedure of the PC algorithm. The algorithm uses all our observed variables to infer a skeleton of a causal graph.

Author: ```suggestion
To begin, it helps to understand the the PC algorithm's general procedure.
It uses all our observed variables to infer a skeleton of a causal graph.
```

Author: See edits in #170.

That is, the algorithm attempts to infer an undirected graph that denotes which variables relate to which other variables, ignoring the directionality of causation between them.
Then, after inferring a skeleton, the algorithm attempts to orient the undirected edges as much as possible.
Optimistically, we end up with a fully directed acyclic graph.
In other cases, we recover a mix of directed and undirected edges, denoting a MEC of graphs with the same independence properties.
In the final case, we recover an undirected causal graph.
This corresponds to the situation where we cannot infer which objects cause which other objects.

> We merely know that given sets of objects relate to each other, not why this relationship exists.

Author: ```suggestion
We merely know that given sets of objects relate to each other, not why these relationships exist.
```

Author: See edits in #170.

With these basics understood, we can now present the results of using the PC algorithm to infer the causal graph for the variables present in Figure \ref{fig:graph-for-testing}.
For reference, the variables (travel time, travel cost, travel distance, number of automobiles in one's household, and the number of licensed drivers in one's household) are all thought to influence the utility of commuting by driving-alone.
We care about the causal relationships between those explanatory variables because intervening on any one of these variables may cause downstream impacts on another explanatory variable.

> To make accurate causal inferences, we need to know and account for these differing pathways of influence on one's utility and mode choice when estimating the causal effect of interventions or treatments.

Author: ```suggestion
As shown in Section \ref{sec:graph-importance}, to make accurate causal inferences, we need to know and account for these differing pathways of influence on one's utility and mode choice.
```

Author: See edits in #170.

In particular, one should think of three kinds of experiments: planned, unplanned, and natural.

Planned experiments are what we typically think of as experiments.
These include A/B tests in technology companies, medical trials for drug certification, randomized controlled trials by economists and social scientists, etc.
Unplanned experiments are instances of true randomization that were not pre-meditated for experimentation.
For instance, technology companies commonly use random allocation (e.g., `np.random.choice` in Python) throughout their code bases, outside of formal experiments.
These are often placeholders or temporary measures, but each such instance is an experiment that allocates some individuals to one treatment and some to another.

Finally, natural experiments are instances where our system is subject to non-random interventions.
Sometimes, these natural experiments take the form of a policy that we apply uniformly across our system.
For example, imagine a country-wide change in immigration policy that affects the supply of employees into the country, when one is analyzing firm success.
In other instances, the interventions affect specific individuals.

> For instance, rounding prices up or down to the nearest dollar allocates people with a ``true price'' of $2.51 to a $3.00 treatment, and it allocates people with a ``true price'' of $2.49 to a $2.00 treatment.

Author: ```suggestion
For instance, rounding prices up or down to the nearest dollar allocates people with a ``true price'' of \$2.51 to a \$3.00 treatment, and it allocates people with a ``true price'' of \$2.49 to a \$2.00 treatment.
```

Author: See edits in #170.
